Add a script to analyze python module dependencies using static analysis

gjoliver opened the pull request:
## Why are these changes needed?

This script can potentially be integrated with CI framework to figure out the exact set of units for python PRs.

## Related issue number

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

simon-mo commented:
quick question, would using [semgrep](https://semgrep.dev/) help reduce the amount of regex and maintenance overhead of this script? or [ast parse](https://docs.python.org/3/library/ast.html#imports) each file and find the `Import` node?

gjoliver commented:
> quick question, would using [semgrep](https://semgrep.dev/) help reduce the amount of regex and maintenance overhead of this script? or [ast parse](https://docs.python.org/3/library/ast.html#imports) each file and find the `Import` node?

I was trying to not install any external libraries, so ast will be awesome I guess.
Give me like a day to play with it, and we will have much better idea.

simon-mo commented:
additionally library should be ok here because we just need them in the CI machine, not users machine

gjoliver commented:
> additionally library should be ok here because we just need them in the CI machine, not users machine

yep. 👌

gjoliver commented:
ast is perfect. this PR is now much simplified and more reliable.
@simon-mo, please take another look. thanks for the suggestion.

simon-mo commented on ci/travis/py_dep_analysis.py:
```diff
@@ -0,0 +1,381 @@
+    return _full_module_path(_bazel_path_to_module_path(dir), fn)
+
+
+def _depends(graph: DepGraph, visited: Dict[int, bool], tid: int, qid: int):
```
need return type annotation

gjoliver replied:
done

simon-mo commented on ci/travis/py_dep_analysis.py:
```diff
@@ -0,0 +1,381 @@
+    return []
+
+
+def test_depends_on_file(graph: DepGraph, test: str, path: str):
```
return type annotation and doc

gjoliver replied:
done

simon-mo reviewed (approved):
Nice! Let's run them in CI. Can you add a BUILD file and run this in one of the jobs in .buildkite/pipeline.yml? Maybe even try to run it in the CI for RLlib? (just print out the dep graph but not using it)

gjoliver commented:
@simon-mo, I added return annotations for all the functions.
and added a dry run of py_dep_analysis in determine_tests_to_run.

however, can you share how and where does determine_tests_to_run run on our CI machine?
py_dep_analysis currently assumes we are running it from ray root dir. so if that's not the case, I need to set the proper base_dir etc.

thanks.

simon-mo commented:
@mwtian FYI. This might be another way of enhancing bazel to do conditional tests.
